content/article/why-i-hate-json-and-you-can-too-.md

```diff
@@ -0,0 +1,177 @@
+
+  {
+    "title"       : "Why I Hate JSON (and you can, too)",
+    "authors"     : ["felipe-gasper"],
+    "date"        : "2019-11-10T21:52:51",
+    "tags"        : [],
+    "draft"       : true,
+    "image"       : "",
+    "thumbnail"   : "",
+    "description" : "Another look at this ubiquitous serialization …",
+    "categories"  : "data"
+  }
+
+[JSON](https://www.json.org/) needs no introduction among most programmers.
+Since its “discovery” by Douglas Crockford in the early 2000s it has become
+one of the most popular data serialization formats yet created, having
+effectively supplanted XML as the web’s “go-to” format for representing
+structured data in APIs.
+
+For all its popularity, though, JSON carries a number of liabilities that
+make it less than ideal in some contexts where it nonetheless enjoys
+widespread use. This article will discuss some of these and propose
+alternative formats that I think deserve your attention.
+
+
+# 1. UTF-8
```

Comment:
I'd disagree that this is it's "chief" drawback. I think a slightly less string wording would avoid a lot of criticism. 

Personally, I'd promote the excessive stringificiation of numbers (which motivated things such as Sereal).

Reply:
Interesting … but how can Sereal solve that problem? Isn’t that just part-and-parcel of encoding into a serialization that’s more precise than the source? Any serialization that distinguishes between numbers and strings will compel a Perl encoder to decide what type suits a given scalar best.

The numbers/strings/booleans problem is annoying, but it’s by no means unsolvable; the non-Perl application just has to know how to typecast (just as a Perl application would have to distinguish between `ne` or `!=` to use on the decoded value).

The UTF-8 problem, on the other hand, seems far more imposing because the serialization literally cannot store anything but UTF-8. There are quite a lot of Windows machines out there whose text editors still output Windows-1252, so a copy-paste from MS Word can yield a text document with, e.g., \x93 and \x94 as single bytes. JSON.pm will screw that up because it coerces it into UTF-8, which means it becomes U+0093 or U+0094, which of course aren’t curly quotes.

Reply:
Sereal uses a binary format for some values so you don't have to unstringify numbers. The booking.com blog posts on the design of Sereal are interesting: https://medium.com/booking-com-development/sereal-a-binary-data-serialization-format-f5ebd6ede507

Reply:
I should also note that any particular drawback doesn't have an inherent "priority" until it shows up in a particular use (same as "best' practice).

If this is the major problem in stuff you've done, tell a story about how it made your life hard.

Reply:
I didn’t mention stringification of numbers in my original post for the reason you cite: any serialization format that distinguishes between strings & numbers will be problematic.

(Incidentally, on our team we’re looking at deploying something like JSON::Schema::Fit to solve this problem.)

I started work last night on fleshing this section out more, with examples, etc.

Reply:
My point isn't about round-tripping serialization: it's that serialization takes time and cycles—even if it could round-trip perfectly.